### ISSUE TITLE
Reduce simulations by using "always better" comparisons

### ISSUE THREAD

**Issue by vn971** [wesnoth/wesnoth#1807]
Currently wesnoth considers all weapons for a battle. There are situations though where one weapon is strictly better than the other. For example, if there are two melee weapons with the same number of strikes and no abilities, one of them is always strictly better than the other (for an exact battle).

By eliminating "always worse" weapons we can reduce the number of simulations we need to calculate a battle. Which in turn should significantly improve battle performance, especially for AI.

The proposal is to sort weapons into categories by the following parameters:
    [range, number of strikes, chance to hit, full abilities list]
prior to a battle. In each category, we can just choose a weapon with the most powerful damage, and ignore all other weapons within the same category.

Note that "first strike" ability may be dropped out for attack moves. "magical", "backstab", "charge" and "maksman" abilities can always be dropped out from abilities list when doing weapon elimilation, only their final damage and chance to hit is what counts when doing categorization.

**Comment by jyrkive** [comment]
I have a different and much more simple idea.

In [battle_context::better_attack()](https://github.com/wesnoth/wesnoth/blob/fbc7652a0cb4cd10e04dfad76203c778efd7a072/src/actions/attack.cpp#L436-L441) it would be possible to detect when the two attacks are otherwise identical (same chance to hit, same retaliation from the opponent, nothing special like poisoning involved) and the only difference is either the **number of hits** or the **damage per hit**. In that case, one attack can be proven to be better than the other without running the whole attack simulation. (Note that calling `get_attacker_combatant()` or `get_defender_combatant()` will run the simulation if it hasn't been run already.)

**Comment by vn971** [comment]
@jyrkive that could be the way to go. It has a downside of actually allowing running an unneeded simulation though. For example, given three attacks a1, a2, a3, we could have a situation where a3 is provably better than a2, but the first comparison we make is a1-versus-a2, which has no such proof, so we will run the full simulation for a2. Pre-sorting avoids this downside.

**Comment by ProditorMagnus** [comment]
Events invalidate the provably better part. Similar #1360 #3923

**Comment by vgaming** [comment]
@ProditorMagnus do they invalidate the approach, or are events just out of reach?

The purpose of this feature request is to speed up AI calculation, pick the best weapon faster. The AI won't trigger events in its simulation anyway, would it?

**Comment by ProditorMagnus** [comment]
Without events the constraints sound better. There are more relevant attributes such as parry, but idea has potential.

**Comment by jyrkive** [comment]
@vgaming Attack simulation is indeed unaware of event interference.

**Comment by Pentarctagon** [comment]
Is this something that can actually be done, given events?

**Comment by soliton-** [comment]
Events don't really play a role here, no way to pick the right weapon based on them. If you have a weapon with 1 damage and one with 10 damage and otherwise the same stats and events somehow make the 1 damage weapon the better one then the AI will simply never choose the better weapon.

Seems fine to disregard the 1 damage weapon in this case but I'm not sure it's going to make a significant difference. Maybe in scenarios/campaigns with units with lots of weapons where this is actually likely to come up.

**Comment by vgaming** [comment]
Is it correct that the *simulation* is aware of events and would "see" them? If that's the case, I'm in for closing the issue.

**Comment by soliton-** [comment]
What do you mean with simulation?

**Comment by vgaming** [comment]
@soliton- I mean the battle prediction system that uses the MK algorithm to predict the approximate battle results.

**Comment by soliton-** [comment]
A little confusing to abbreviate monte carlo with MK but I assume that's what you mean. It's not a simulation in that sense. It's just a different way to calculate the probabilities.

**Comment by vgaming** [comment]
@soliton- you're right, fixed the type. I agree it's just a way to measure probabilities. I'm asking if that system is aware of events. If it is not, then it won't reveal anything that other naive systems would see. So that would be a point of eliminating useless tasks for the battle simulation engine as per the proposal in this issue.
